Update lakefile with scripts for import generation

0art0 opened the pull request:
This pull request introduces two new scripts to the `lakefile` of `mathlib4`:
- `import_all`: Generate top-level files importing all the files in the libraries of the repository.
- `import_all?`: Check whether the top-level files are up-to-date, which may be useful for CI (this is modelled on [the `lakefile` of `Yatima`](https://github.com/lurk-lab/yatima/blob/main/lakefile.lean#L133-L144), suggested by @arthurpaulino .)

Note that the scripts in this `lakefile` are not inherited by downstream projects with a `mathlib4` dependency.

The scripts can be run by typing `lake run import_all` and `lake run import_all?` respectively.
 
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

0art0 committed:
Update lakefile with scripts for import generation

semorrison commented:
@0art0, just linking to @tydeu's messages so we can track that these issues are addressed here.

https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Lake.20script.20for.20automatically.20generating.20files.20with.20imports/near/370935304 (`getCurrentPackage` will break when run from another directory)

0art0 committed:
Generalise `mk_import` script

0art0 commented on lakefile.lean:
```diff
@@ -53,39 +53,41 @@ section Scripts
+def importsForLib (dir : FilePath) (root : Name) : IO String := do
+  let files ← FilePath.walkDir (dir / root.toString) >>= Array.filterM (not <$> ·.isDir)
+  let filePathToImport (fp : FilePath) : String := fp.toString
+    |>.drop dir.toString.length.succ
```
Is there a more robust way to achieve this? I would like to drop the prefix corresponding to the directory `dir` and retain the remaining portion of the relative file path.

tydeu replied:
You cloud use what Lake does for globs (i.e., [`forEachModuleIn`](https://github.com/leanprover/lake/blob/e9745480069793e8276158283bddff968426efa3/Lake/Config/Glob.lean#L26-L35)) or just emulate it.

0art0 replied:
Thanks. I have tried a different approach by creating a function `System.FilePath.makeRelativeTo` that performs the necessary check before modifying the path.

0art0 commented:
I have updated the script according to @tydeu's suggestions [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Lake.20script.20for.20automatically.20generating.20files.20with.20imports).

0art0 committed:
More robust implementation of relative `FilePath`s

0art0 committed:
Fix issue

0art0 commented on lakefile.lean:
```diff
@@ -48,3 +48,53 @@ lean_lib Archive where
+
+open System
+
+def System.FilePath.makeRelativeTo (dir file : FilePath) : Option FilePath := do
```
Should this function be moved elsewhere?

tydeu replied:
Personally, I would suggest using `readDir` instead of `walkDir` so you can avoid this post-processing.

0art0 committed:
Modification (not working)

0art0 committed:
Fix `import_all` script

0art0 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into import-scripts

0art0 commented:
Apologies for the delay, there were a few subtle issues in the code (which are now fixed). I have incorporated @tydeu's suggestion of using a variant of `forEachModuleIn` to avoid ad-hoc processing of file names. 

Please let me know if I should run `lake run import_all` and push the results up before this PR is merged. As I mentioned earlier, the order of the imports in `Mathlib.lean` and related top-level files might change - the current order is alphabetical, but the script in this PR orders them according to the depth-first traversal of the file directory.

0art0 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into import-scripts

0art0 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into import-scripts

0art0 committed:
Add newline

0art0 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into import-scripts

0art0 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into import-scripts

semorrison commented:
I think `Mathlib.lean` should remain alphabetical.

0art0 committed:
Tweak function `importsForLib`

0art0 committed:
Alphabetical sorting of filenames

0art0 committed:
`includeDirs?` option in `moduleNamesIn`

0art0 commented:
I have now made the imports alphabetically-ordered.

semorrison commented:
I guess if we are going to switch to this, we should replace the `check_imported` section of the `.github/workflows/*.yml` with an invocation of this script?

That would require further modification to allow building `Mathlib/Tactic.lean`.

semorrison commented:
I'm slightly uncomfortable adding any code at all to the lakefile. This adds to the elaboration time of every `lake build`.

For example, on my system `time (for i in {1..100}; do lake env echo; done)` (i.e. elaborate the lakefile, but do nothing, 100 times) takes 223s on this branch, but on 185s on master.

That means that the mere presence of these scripts in the lakefile is adding about `0.4s` to every single invocation of `lake`, which is unfortunately not viable. Pinging @tydeu in case I'm missing something here.

We could move these scripts into the codebase (e.g. like ImportGraph) for now.

0art0 commented:
Would moving these scripts to one of the dependencies, say `Std`, be an option? Lake scripts can be invoked in downstream projects (in this case, it will be `lake run std/import_all` instead of `lake run import_all`), as Mac mentioned in a different conversation on Zulip.

tydeu commented:
@0art0 Not at the moment, Lake elaborates all configuration files (both the root and dependent packages) on each startup. This is something I hope to address soon.

0art0 commented:
Regarding building `Mathlib/Tactic.lean`, is there a way to make the `Mathlib/Tactic` folder a package in itself? I tried adding variations of
```lean
lean_lib Tactic where
  roots := #[ `«Mathlib/Tactic» ]
```
to the `lakefile`, but haven't had any success yet. If this works, I think `Mathlib/Tactic.lean` will be built automatically by the script each time.

As a follow-up question: even if there is a way to add `Mathlib/Tactic` as a package to the `lakefile`, is this recommended practice?

grunweg commented:
It seems this PR is superseded by #11849 and #11874 now (which implement the same idea, outside of the lakefile).
In other words, the effect of this is coming to mathlib. Thank you for (perhaps) starting this conversation.
